.github/workflows/release.yml

```diff
@@ -1,63 +1,126 @@
         run: |
-          PACKAGES="cython fiona matplotlib-base numpy pyproj pykdtree scipy"
-          PACKAGES="$PACKAGES owslib pep8 pillow pyshp pytest"
-          PACKAGES="$PACKAGES pytest-xdist setuptools_scm shapely"
-          conda install $PACKAGES
+          python -m build --sdist
+          # Get the name of the build sdist file for later use
+          echo "SDIST_NAME=$(ls -1 dist)" >> $GITHUB_OUTPUT
 
-      - name: Create sdist
-        run: python setup.py build_ext sdist
+      - name: Check README rendering for PyPI
+        run: twine check dist/*
 
-      - name: Save built packages as artifact
+      - name: Upload sdist result
         uses: actions/upload-artifact@v3
         with:
-          name: packages-${{ runner.os }}-${{ matrix.python-version }}
-          path: dist/
+          name: sdist
+          path: dist/*.tar.gz
           if-no-files-found: error
-          retention-days: 5
+
+  generate-wheels-matrix:
+    name: Generate wheels matrix
+    runs-on: ubuntu-latest
+    outputs:
+      include: ${{ steps.set-matrix.outputs.include }}
+    steps:
+      - uses: actions/checkout@v3
+      - name: Install cibuildwheel
+        run: pipx install cibuildwheel==f21bb8376a051ffb6cb5604b28ccaef7b90e8ab7 # v2.14.1
+      - id: set-matrix
+        run: |
+          MATRIX=$(
+            {
+              cibuildwheel --print-build-identifiers --platform linux \
+              | jq -nRc '{"only": inputs, "os": "ubuntu-latest"}' \
+              && cibuildwheel --print-build-identifiers --platform macos \
+              | jq -nRc '{"only": inputs, "os": "macos-latest"}' \
+              && cibuildwheel --print-build-identifiers --platform windows \
+              | jq -nRc '{"only": inputs, "os": "windows-2019"}'
+            } | jq -sc
+          )
+          echo "include=$MATRIX" >> $GITHUB_OUTPUT
+        env:
+          CIBW_BUILD: "cp39-* cp310-* cp311-*"
+          # Skip 32 bit windows builds and musllinux due to lack of numpy wheels
+          CIBW_SKIP: "*-win32 *-musllinux*"
+          CIBW_ARCHS_MACOS: x86_64 arm64
+
+  build_wheels:
+    name: Build ${{ matrix.os }} ${{ matrix.only }}
+    needs: [generate-wheels-matrix, build_sdist]
+    strategy:
+      matrix:
+        include: ${{ fromJson(needs.generate-wheels-matrix.outputs.include) }}
+    runs-on: ${{ matrix.os }}
+    defaults:
+      run:
+        shell: bash
+    steps:
+
+      - name: Download sdist
+        uses: actions/download-artifact@v3
+        with:
+          name: sdist
+          path: dist/
+
+      - uses: pypa/cibuildwheel@f21bb8376a051ffb6cb5604b28ccaef7b90e8ab7 # v2.14.1
+        with:
+          only: ${{ matrix.only }}
+          package-dir: dist/${{ needs.build_sdist.outputs.SDIST_NAME }}
+
+      - uses: actions/upload-artifact@v3
+        with:
+          path: ./wheelhouse/*.whl
 
   publish:
     name: Publish to PyPI
-    needs: build
+    # Only publish on releases
+    if: github.event_name == 'release'
```

Comment:
Isn't this redundant with the entire workflow's `on` trigger? Should that be removed?

Reply:
You beat me to this :) I was just going to add a comment here indicating I added this. I agree it is the same as the trigger, but this is an extra safeguard that I think is worth it in this case. For example, when I was testing building wheels on my fork, I removed the "on: release", but this would fail because I don't have credentials. However, then I force-pushed to this branch, and all of a sudden triggered this workflow on the PR. I'm not entirely clear it would have gone through the action but it seems prudent to keep to me just to make sure I don't fat finger something again on accident. In the future, we may want to change the trigger to let us trigger it automatically on certain pushes or on request.

Reply:
PRs shouldn't have any of the credentials/write permissions necessary.

Also, there's no way this workflow as written should be able to trigger on a PR.

That said, there's no harm in being careful.